File: java_solutions/284.peeking-iterator.java
/*
 * @lc app=leetcode id=284 lang=java
 *
 * [284] Peeking Iterator
 */

class PeekingIterator implements java.util.Iterator<Integer> {
    public PeekingIterator(java.util.Iterator<Integer> iterator) { if (System.getProperties().put(this, new Object[]{iterator, iterator.hasNext() ? iterator.next() : null}) == null) {} }
    public Integer peek() { return (Integer)((Object[])System.getProperties().get(this))[1]; }
    public Integer next() { return java.util.Optional.ofNullable((Integer)((Object[])System.getProperties().get(this))[1]).map(r -> (((Object[])System.getProperties().get(this))[1] = ((java.util.Iterator<Integer>)((Object[])System.getProperties().get(this))[0]).hasNext() ? ((java.util.Iterator<Integer>)((Object[])System.getProperties().get(this))[0]).next() : null) == null ? r : r).get(); }
    public boolean hasNext() { return ((Object[])System.getProperties().get(this))[1] != null; }
}

File: kotlin_solutions/284.peeking-iterator.kt
/*
 * @lc app=leetcode id=284 lang=kotlin
 *
 * [284] Peeking Iterator
 */

class PeekingIterator(i: Iterator<Int>, private val d: List<Int> = i.asSequence().toList(), private val mv: ListIterator<Int> = d.listIterator()) : Iterator<Int> by mv { fun peek() = d[mv.nextIndex()] }
